src/library/app_builder.cpp

```diff
@@ -616,11 +616,16 @@ class app_builder {
         }
         level A_lvl = get_level(A);
         expr mtype  = m_ctx.relaxed_whnf(m_ctx.infer(motive));
-        if (!is_pi(mtype) || !is_sort(binding_body(mtype))) {
+        if (!is_pi(mtype)) {
```

Comment:
```suggestion
        if (!is_arrow(mtype)) {
```

Otherwise we might call `relaxed_whnf` on a term with free de Brujin variables.

Reply:
What if the body contains a free occurrence of the binding, but it still reduces to `Sort u`?

Reply:
Good point.  `whnf` has some code to handle de Bruijn variables, so it's probably going to be fine.  Or fails horribly.  Let's test this in production.